flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/CreditBasedInputBufferPoolMetric.java

```diff
@@ -0,0 +1,136 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.metrics;
+
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.runtime.io.network.partition.consumer.InputChannel;
+import org.apache.flink.runtime.io.network.partition.consumer.RemoteInputChannel;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate;
+
+/**
+ * Gauge metric measuring the input buffer pool usage gauge for {@link SingleInputGate}s under credit based mode.
+ */
+public class CreditBasedInputBufferPoolMetric {
+
+	/**
+	 * Gauge metric measuring the floating buffer usage for {@link SingleInputGate}.
+	 */
+	public static class FloatingBufferPoolUsage implements Gauge<Float> {
+
+		public SingleInputGate[] inputGates;
+
+		public FloatingBufferPoolUsage(SingleInputGate[] inputGates) {
+			this.inputGates = inputGates;
+		}
+
+		@Override
+		public Float getValue() {
+			int leftFloatingBuffer = 0;
+			int requestedFloatingBuffer = 0;
+
+			for (SingleInputGate inputGate : inputGates) {
+				requestedFloatingBuffer += inputGate.getBufferPool().bestEffortGetNumOfUsedBuffers();
+				for (InputChannel ic : inputGate.getInputChannels().values()) {
+					if (ic instanceof RemoteInputChannel) {
+						leftFloatingBuffer += ((RemoteInputChannel) ic).unsafeGetFloatingBufferLeft();
+					}
+				}
+			}
+
+			if (requestedFloatingBuffer == 0) {
+				return 0.0f;
+			} else {
+				return 1 - ((float) leftFloatingBuffer) / requestedFloatingBuffer;
+			}
+		}
+	}
+
+	/**
+	 * Gauge metric measuring the exclusive buffer usage for {@link SingleInputGate}.
+	 */
+	public static class ExclusiveBufferPoolUsage implements Gauge<Float> {
+
+		public SingleInputGate[] inputGates;
+
+		public ExclusiveBufferPoolUsage(SingleInputGate[] inputGates) {
+			this.inputGates = inputGates;
+		}
+
+		@Override
+		public Float getValue() {
+			int usedExclusiveBuffer = 0;
+			int requestedExclusiveBuffer = 0;
+
+			for (SingleInputGate inputGate : inputGates) {
+				for (InputChannel ic : inputGate.getInputChannels().values()) {
+					if (ic instanceof RemoteInputChannel) {
+						requestedExclusiveBuffer += ((RemoteInputChannel) ic).getInitialCredit();
+						usedExclusiveBuffer += ((RemoteInputChannel) ic).unsafeGetExclusiveBufferUsed();
+					}
+				}
+			}
+
+			if (requestedExclusiveBuffer != 0) {
+				return ((float) usedExclusiveBuffer) / requestedExclusiveBuffer;
+			} else {
+				return 0.0f;
+			}
+		}
+	}
+
+	/**
+	 * Gauge metric measuring the overall buffer usage for {@link SingleInputGate} under credit based mode.
+	 */
+	public static class InputBufferPoolUsageGauge implements Gauge<Float> {
+
+		public SingleInputGate[] inputGates;
+
+		public InputBufferPoolUsageGauge(SingleInputGate[] inputGates) {
+			this.inputGates = inputGates;
+		}
+
+		@Override
+		public Float getValue() {
```

Comment:
We might extract the logics of above `exclusiveBuffersUsage and `floatingBuffersUsage into separate methods, then these methods could be reused here to get final results directly, to avoid below detail logics again.